src/Polly.Core/Timeout/OnTimeoutEvent.cs

```diff
@@ -0,0 +1,78 @@
+namespace Polly.Timeout;
+
+/// <summary>
+/// An event that is raised when a timeout occurs.
+/// </summary>
+/// <remarks>This class supports registering multiple on-timeout callbacks.</remarks>
+public sealed class OnTimeoutEvent
+{
+    private readonly List<Func<OnTimeoutArguments, ValueTask>> _events = new();
+
+    /// <summary>
+    /// Adds an asynchronous on-timeout callback.
+    /// </summary>
+    /// <param name="onTimeout">The specified callback.</param>
+    /// <returns>This instance.</returns>
+    public OnTimeoutEvent Add(Func<OnTimeoutArguments, ValueTask> onTimeout)
+    {
+        Guard.NotNull(onTimeout);
+
+        _events.Add(onTimeout);
+        return this;
+    }
+
+    /// <summary>
+    /// Adds a synchronous on-timeout callback.
+    /// </summary>
+    /// <param name="onTimeout">The specified callback.</param>
+    /// <returns>This instance.</returns>
+    public OnTimeoutEvent Add(Action<OnTimeoutArguments> onTimeout)
+    {
+        Guard.NotNull(onTimeout);
+
+        _events.Add(args =>
+        {
+            onTimeout(args);
+            return default;
+        });
+        return this;
+    }
+
+    /// <summary>
+    /// Adds a synchronous on-timeout callback.
+    /// </summary>
+    /// <param name="onTimeout">The specified callback.</param>
+    /// <returns>This instance.</returns>
+    public OnTimeoutEvent Add(Action onTimeout)
+    {
+        Guard.NotNull(onTimeout);
+
+        _events.Add(_ =>
+        {
+            onTimeout();
+            return default;
+        });
+        return this;
+    }
+
+    internal Func<OnTimeoutArguments, ValueTask> CreateHandler()
+    {
+        return _events.Count switch
+        {
+            0 => _ => default,
```

Comment:
You might want to check whether this allocates a delegate every time through or not. If it does allocate every time through, then you can just use a static field to hold an instance.

Alternatively, if the 0 case is common, it might be more efficient to return null here, and to have the call site do null checks before invoking the delegate. Calling into a delegate will be more expensive than the null check. If 0 is rare, then this approach is better.

Reply:
I have added the `EmptyEvent` static field that's used when there are no callbacks. This also improves the readability. 

> Calling into a delegate will be more expensive than the null check. If 0 is rare, then this approach is better.

Only focusing on core functionality now, we can do full optimalization pass later once we have the features in. Wdyt?

Reply:
> Only focusing on core functionality now, we can do full optimalization pass later once we have the features in. Wdyt?

I personally find it hard to come back and find/fix this kind of small stuff. But if you prefer doing the micro-optimizations later on, that also definitely works.

Reply:
You are right here, just making the result callback nullable and returning null if no events are registered is simple speeds-up the code with minimal effort.